pkg/detectors/trufflehogenterprisescanner/trufflehogenterprisescanner.go

```diff
@@ -0,0 +1,117 @@
+package trufflehogenterprisescanner
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors/trufflehogenterprisescanner/scannerpb"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/detectorspb"
+)
+
+type Scanner struct{}
+
+// Ensure the Scanner satisfies the interface at compile time.
+var _ detectors.Detector = (*Scanner)(nil)
+
+var (
+	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
+	keyPat     = regexp.MustCompile(`\bthog-agent-[0-9a-f]{32}\b`)
+	groupPat   = regexp.MustCompile(`\bthog-scanner-[a-zA-Z]+\b`)
+	addressPat = regexp.MustCompile(`\b[a-z]+-[a-z]+-[a-z]+.[a-z]+\.[a-z][0-9]\.[a-z]+\.trufflehog\.org:[0-9]{4}\b`)
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"thog"}
+}
+
+// FromData will find and optionally verify TrufflehogEnterpriseScanner secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	keyMatches := keyPat.FindAllStringSubmatch(dataStr, -1)
+	groupMatches := groupPat.FindAllStringSubmatch(dataStr, -1)
+	addressMatches := addressPat.FindAllStringSubmatch(dataStr, -1)
+
+	for _, keyMatch := range keyMatches {
+		if len(keyMatch) != 1 {
+			continue
+		}
+		resKeyMatch := strings.TrimSpace(keyMatch[0])
+
+		for _, groupMatch := range groupMatches {
+			if len(groupMatch) != 1 {
+				continue
+			}
+			resGroupMatch := strings.TrimSpace(groupMatch[0])
+
+			for _, addressMatch := range addressMatches {
+				if len(addressMatch) != 1 {
+					continue
+				}
+
+				resAddressMatch := strings.TrimSpace(addressMatch[0])
+
+				s1 := detectors.Result{
+					DetectorType: detectorspb.DetectorType_TrufflehogEnterpriseScanner,
+					Raw:          []byte(resKeyMatch),
+				}
+
+				if verify {
+					err := grpcHeartbeat(resAddressMatch, resGroupMatch, resKeyMatch)
+					if err == nil {
+						s1.Verified = true
+					}
+				}
+				results = append(results, s1)
+			}
+		}
+	}
+	return results, nil
+
+}
+
+func (s Scanner) Type() detectorspb.DetectorType {
+	return detectorspb.DetectorType_TrufflehogEnterpriseScanner
+}
+
+func grpcHeartbeat(address, group, token string) error {
+	ctx := context.Background()
+	ctx = metadata.NewOutgoingContext(ctx,
+		metadata.New(map[string]string{
+			"name":          strings.TrimSpace(group),
+			"authorization": strings.TrimSpace(token),
+		}))
+
+	systemRoots, err := x509.SystemCertPool()
+	if err != nil {
+		panic(errors.Wrap(err, "cannot load root CA certs"))
```

Comment:
Is this a normal way of dealing with failures during verification?

Reply:
@dustin-decker and i had paired on this grpcHeartbeat and come up with this approach. is there an alternative you'd recommend?

Reply:
i mean panicking. is the thinking here that without the root certs nothing is going to work anyway?

Reply:
that is my understanding

Reply:
I think instead of panicking here, we should just return the error. That way we still record the cred as unverified since it made it through keyword and regex matching. If we panic here there is a chance the entire application will crash mid scan. We don't really want the application to crash b/c one detector failed to verify a secret.

Actually, I do believe we have a recover mechanism for panics, but nonetheless, we likely just want to return the error here and treat it as an unverified result.

Reply:
I think Ahrav is right, this error should be returned. Worth pointing out that currently the errors are swallowed (this will soon change with Cody's work), and that without the root CA store all verification over TLS will fail.

A future improvement we could consider is embedding our own CertPool to fall back to in the default HTTP client if there is no root CA store present. And it's pretty easy to end up in that situation with Docker images.